pkg/registry/uploader.go

```diff
@@ -4,162 +4,75 @@
-	if err != nil {
-		return i.refDigestStr, newImageAccessError(i.refTag.String(), err)
-	}
-
-	return i.refDigestStr, err
+	return d.Relocator.Relocate(keychain, image, dstImgRefStr)
 }
 
-type uploadImageInfo struct {
-	image        v1.Image
-	refTag       name.Reference
-	refDigestStr string
-	size         int64
-}
-
-type uploadCfg struct {
-	imgInfo         uploadImageInfo
-	srcTarPath      string
-	imgWriteOptions []remote.Option
-}
-
-func getImageUploadCfg(imgRefStr, srcPath string, tlsCfg TLSConfig) (uploadCfg, error) {
-	var cfg uploadCfg
-
-	transport, err := tlsCfg.Transport()
-	if err != nil {
-		return cfg, err
-	}
-
+func writeLocalPathToImage(path string) (v1.Image, error) {
 	var srcTarPath string
-	if archive.IsZip(srcPath) {
-		srcTarPath, err = archive.ZipToTar(srcPath)
-	} else {
-		srcTarPath, err = archive.CreateTar(srcPath)
-	}
+	var err error
 
-	if err != nil {
-		return cfg, err
-	}
+	srcTarPath, err = readPathToTar(path)
+	defer os.Remove(srcTarPath)
 
-	info, err := getImageInfo(imgRefStr, srcTarPath)
 	if err != nil {
-		return cfg, err
+		return nil, err
 	}
 
-	cfg = uploadCfg{
-		imgInfo:    info,
-		srcTarPath: srcTarPath,
-		imgWriteOptions: []remote.Option{
-			remote.WithAuthFromKeychain(authn.DefaultKeychain),
-			remote.WithTransport(transport),
-		},
-	}
-	return cfg, err
-}
-
-func getImageInfo(imgRefStr, tarPath string) (uploadImageInfo, error) {
-	var info uploadImageInfo
-
-	image, err := getImageFromSrcTar(tarPath)
-	if err != nil {
-		return info, err
-	}
-
-	tagStr := fmt.Sprint(time.Now().UnixNano())
-	refTag, err := name.ParseReference(fmt.Sprintf("%s:%s", imgRefStr, tagStr))
+	image, err := random.Image(0, 0)
 	if err != nil {
-		return info, err
+		return image, err
 	}
 
-	digest, err := image.Digest()
+	layer, err := tarball.LayerFromFile(srcTarPath)
 	if err != nil {
-		return info, err
+		return image, err
 	}
 
-	size, err := imageSize(image)
+	image, err = mutate.AppendLayers(image, layer)
 	if err != nil {
-		return info, errors.WithStack(err)
+		return image, err
 	}
 
-	info = uploadImageInfo{
-		image:        image,
-		refTag:       refTag,
-		refDigestStr: fmt.Sprintf("%s@%s", imgRefStr, digest),
-		size:         size,
-	}
-	return info, err
+	return image, nil
 }
 
-func getImageFromSrcTar(tarFilepath string) (v1.Image, error) {
-	image, err := random.Image(0, 0)
-	if err != nil {
-		return image, err
-	}
-
-	layer, err := tarball.LayerFromFile(tarFilepath)
+func readPathToTar(path string) (string, error) {
+	fi, err := os.Stat(path)
 	if err != nil {
-		return image, err
+		return "", err
 	}
 
-	image, err = mutate.AppendLayers(image, layer)
-	if err != nil {
-		return image, errors.Wrap(err, "adding layer")
+	if !fi.IsDir() && archive.IsZip(path) {
+		return archive.ZipToTar(path)
+	} else if !fi.IsDir() {
+		return "", errors.New("local path must be a directory or zip")
 	}
 
-	return image, nil
+	return archive.CreateTar(path)
```

Comment:
This writes a tar to disk that we need to remove. A future refactoring could instead provide a ReadCloser.